Shallow fusion & LODR documentation (#1142)
k2-fsa/icefall: docs/source/decoding-with-langugage-models/LODR.rst

.. _LODR:

LODR for RNN Transducer
=======================

As a type of E2E model, neural transducers are usually considered as having an internal 
language model, which learns the language level information on the training corpus. 
In real-life scenario, there is often a mismatch between the training corpus and the target corpus space. 
This mismatch can be a problem when decoding for neural transducer models with language models as its internal
language can act "against" the external LM. In this tutorial, we show how to use
`Low-order Density Ratio <https://arxiv.org/abs/2203.16776>`_ to alleviate this effect to further improve the performance
of langugae model integration. 

.. note::

    This tutorial is based on the recipe 
    `pruned_transducer_stateless7_streaming <https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/pruned_transducer_stateless7_streaming>`_,
    which is a streaming transducer model trained on `LibriSpeech`_. 

>     However, you can easily apply shallow fusion to other recipes.

Reviewer: ```suggestion
    However, you can easily apply LODR to other recipes.
```

For illustration purpose, we will use a pre-trained ASR model from this `link <https://huggingface.co/Zengwei/icefall-asr-librispeech-pruned-transducer-stateless7-streaming-2022-12-29>`_.
If you want to train your model from scratch, please have a look at :ref:`non_streaming_librispeech_pruned_transducer_stateless`.
The testing scenario here is intra-domain (we decode the model trained on `LibriSpeech`_ on `LibriSpeech`_ testing sets).

As the initial step, let's download the pre-trained model.

.. code-block:: bash

    $ git lfs install
    $ git clone https://huggingface.co/Zengwei/icefall-asr-librispeech-pruned-transducer-stateless7-streaming-2022-12-29
    $ pushd icefall-asr-librispeech-pruned-transducer-stateless7-streaming-2022-12-29/exp
    $ ln -s pretrained.pt epoch-99.pt # create a symbolic link so that the checkpoint can be loaded

To test the model, let's have a look at the decoding results **without** using LM. This can be done via the following command:

.. code-block:: bash

    $ exp_dir=./icefall-asr-librispeech-pruned-transducer-stateless7-streaming-2022-12-29/exp/
    $ ./pruned_transducer_stateless7_streaming/decode.py \

>         --epoch 30 \

Reviewer: Since this example is using the downloaded model and we have already created `epoch-99.pt`, please use
```bash
--epoch 99 \
--avg 1 \
--use-averaged-model 0 \
```
so that users can reproduce the results.

Author: Sorry, will change it.

    $ pushd data/lang_bpe_500
    $ ln -s ../../librispeech_bigram/2gram.fst.txt .
    $ popd

Then, we perform LODR decoding by setting ``--decoding-method`` to ``modified_beam_search_lm_LODR``:

.. code-block:: bash
    
    $ exp_dir=./icefall-asr-librispeech-pruned-transducer-stateless7-streaming-2022-12-29/exp
    $ lm_dir=./icefall-librispeech-rnn-lm/exp
    $ lm_scale=0.42
    $ LODR_scale=-0.24
    $ ./pruned_transducer_stateless7_streaming/decode.py \
        --epoch 99 \
        --avg 1 \
        --use-averaged-model False \
        --beam-size 4 \
        --exp-dir $exp_dir \
        --max-duration 600 \
        --decode-chunk-len 32 \
        --decoding-method modified_beam_search_lm_LODR \
        --bpe-model ./icefall-asr-librispeech-pruned-transducer-stateless7-streaming-2022-12-29/data/lang_bpe_500/bpe.model
        --use-shallow-fusion 1 \
        --lm-type rnn \
        --lm-exp-dir $lm_dir \
        --lm-epoch 99 \
        --lm-scale $lm_scale \
        --lm-avg 1 \
        --rnn-lm-embedding-dim 2048 \
        --rnn-lm-hidden-dim 2048 \
        --rnn-lm-num-layers 3 \
        --lm-vocab-size 500 \
        --tokens-ngram 2 \
        --ngram-lm-scale $LODR_scale

> There are two extra arguments need to be given when doing LODR. ``--tokens-ngram`` specifies the order of n-gram. As we

Reviewer: ```suggestion
There are two extra arguments that need to be given when doing LODR. ``--tokens-ngram`` specifies the order of n-gram. As we
```

are using a bi-gram, we set it to 2. ``--ngram-lm-scale`` is the scale of the bi-gram, it should be a negative number
as we are subtracting the bi-gram's score during decoding.

The decoding results obtained with the above command are shown below:

.. code-block:: text

    $ For test-clean, WER of different settings are:
    $ beam_size_4	2.61	best for test-clean
    $ For test-other, WER of different settings are:
    $ beam_size_4	6.74	best for test-other

> Recall that the lowest WER we obtained in :ref:`shallow_fusion` with beam size of 4 is 2.77/7.08, LODR

Reviewer: ```suggestion
Recall that the lowest WER we obtained in :ref:`shallow_fusion` with beam size of 4 is ``2.77/7.08``, LODR
```

.. _LODR:

LODR for RNN Transducer
=======================

As a type of E2E model, neural transducers are usually considered as having an internal 
language model, which learns the language level information on the training corpus. 
In real-life scenario, there is often a mismatch between the training corpus and the target corpus space. 
This mismatch can be a problem when decoding for neural transducer models with language models as its internal
language can act "against" the external LM. In this tutorial, we show how to use
`Low-order Density Ratio <https://arxiv.org/abs/2203.16776>`_ to alleviate this effect to further improve the performance
of langugae model integration. 

.. note::

    This tutorial is based on the recipe 
    `pruned_transducer_stateless7_streaming <https://github.com/k2-fsa/icefall/tree/master/egs/librispeech/ASR/pruned_transducer_stateless7_streaming>`_,
    which is a streaming transducer model trained on `LibriSpeech`_. 
    However, you can easily apply shallow fusion to other recipes.
    If you encounter any problems, please open an issue here `icefall <https://github.com/k2-fsa/icefall/issues>`_.

.. note::

>     For simplicity, the training and testing corpus in this tutorial is the same (`LibriSpeech`_). However, 

Reviewer: ```suggestion
    For simplicity, the training and testing corpus in this tutorial are the same (`LibriSpeech`_). However, 
```

corpus (source domain) and the testing corpus (target domain). Assuming that the source domain and the test domain
are acoustically similar, DR derives the following formular for decoding with Bayes' theorem:

.. math::

    \text{score}\left(y_u|\mathit{x},y\right) = 
    \log p\left(y_u|\mathit{x},y_{1:u-1}\right) + 
    \lambda_1 \log p_{\text{Target LM}}\left(y_u|\mathit{x},y_{1:u-1}\right) - 
    \lambda_2 \log p_{\text{Source LM}}\left(y_u|\mathit{x},y_{1:u-1}\right)

where :math:`\lambda_1` and :math:`\lambda_2` are the LM score for target domain and source domain respectively. 
Here, the source domain LM is trained on the training corpus. The only difference in the above formular compared to 
shallow fusion is the subtraction of the source domain LM.

Some works treat the predictor and the joiner of the neural transducer as its internal LM. However, the LM is 
considered to be weak and can only capture low-level language information. Therefore, `LODR <https://arxiv.org/abs/2203.16776>`__ proposed to use
a low-order n-gram LM as an approximation of the ILM of the neural transducer. This leads to the following formula

> during decoding for RNNT model:

Reviewer: ```suggestion
during decoding for transducer model:
```

.. math::

    \text{score}\left(y_u|\mathit{x},y\right) = 
    \log p_{rnnt}\left(y_u|\mathit{x},y_{1:u-1}\right) + 
    \lambda_1 \log p_{\text{Target LM}}\left(y_u|\mathit{x},y_{1:u-1}\right) - 
    \lambda_2 \log p_{\text{bi-gram}}\left(y_u|\mathit{x},y_{1:u-1}\right)

In LODR, an additional bi-gram LM estimated on the training corpus is required apart from the neural LM. Comared to DR, 
the only difference lies in the choice of source domain LM. According to the original `paper <https://arxiv.org/abs/2203.16776>`_,

> LODR achieves similar performance compared DR. As a bi-gram is much faster to evaluate, LODR

Reviewer: ```suggestion
LODR achieves similar performance compared to DR. As a bi-gram is much faster to evaluate, LODR
```

is usually much faster. Note that although DR/LODR is originally proposed to address the domain
mismatch between training and testing, it still achieves very good results on intra-domain evaluation.

Now, we will show you how to use LODR in ``icefall``.
For illustration purpose, we will use a pre-trained ASR model from this `link <https://huggingface.co/Zengwei/icefall-asr-librispeech-pruned-transducer-stateless7-streaming-2022-12-29>`_.
If you want to train your model from scratch, please have a look at :ref:`non_streaming_librispeech_pruned_transducer_stateless`.
The testing scenario here is intra-domain.

As the initial step, let's download the pre-trained model.

.. code-block:: bash

>     $ git lfs install

Reviewer: please switch to the download method described in
https://k2-fsa.github.io/sherpa/cpp/pretrained_models/online_transducer.html#icefall-asr-librispeech-pruned-transducer-stateless7-streaming-2022-12-29

otherwise, it makes `git clone` download all the files from the repo, including the unnecessary very large file(s).

Author: Sure, will do
